velox/exec/HashBuild.cpp

```diff
@@ -178,12 +278,315 @@ void HashBuild::addInput(RowVectorPtr input) {
+    // Enough free rows for input rows and enough variable length free
+    // space for the flat size of the whole vector. If outOfLineBytes
+    // is 0 there is no need for variable length space.
+    return true;
+  }
+
+  // If there is variable length data we take the flat size of the
+  // input as a cap on the new variable length data needed.
+  const auto increment =
+      rows->sizeIncrement(input->size(), outOfLineBytes ? flatBytes : 0);
+
+  auto tracker = CHECK_NOTNULL(mappedMemory_->tracker());
+  // There must be at least 2x the increments in reservation.
+  if (tracker->getAvailableReservation() > 2 * increment) {
+    return true;
+  }
+
+  // Check if we can increase reservation. The increment is the larger of
+  // twice the maximum increment from this input and
+  // 'spillableReservationGrowthPct_' of the current reservation.
+  auto targetIncrement = std::max<int64_t>(
+      increment * 2,
+      tracker->getCurrentUserBytes() *
+          spillConfig()->spillableReservationGrowthPct / 100);
+  if (tracker->maybeReserve(targetIncrement)) {
+    return true;
+  }
+  numSpillRows_ = std::max<int64_t>(
+      1, targetIncrement / (rows->fixedRowSize() + outOfLineBytesPerRow));
+  numSpillBytes_ = numSpillRows_ * outOfLineBytesPerRow;
+  return false;
+}
+
+void HashBuild::spillInput(RowVectorPtr& input) {
+  VELOX_CHECK_EQ(input->size(), activeRows_.size());
+
+  if (!spillEnabled() || !spiller_->isAnySpilled() ||
+      !activeRows_.hasSelections()) {
+    return;
+  }
+
+  const auto numInput = input->size();
+  prepareInputIndicesBuffers(numInput, spiller_->spilledPartitionSet());
+  computeSpillPartitions(input);
+
+  vector_size_t numSpillInputs = 0;
+  for (auto row = 0; row < numInput; ++row) {
+    const auto partition = spillPartitions_[row];
+    if (FOLLY_UNLIKELY(!activeRows_.isValid(row))) {
+      continue;
+    }
+    if (!spiller_->isSpilled(partition)) {
+      continue;
+    }
+    activeRows_.setValid(row, false);
+    ++numSpillInputs;
+    spillInputIndicesBuffers_[partition]
+        ->asMutable<vector_size_t>()[numSpillInputs_[partition]++] = row;
+  }
+  if (numSpillInputs == 0) {
+    return;
+  }
+
+  for (uint32_t partition = 0; partition < numSpillInputs_.size();
+       ++partition) {
+    const int numSpillInputs = numSpillInputs_[partition];
+    if (numSpillInputs == 0) {
+      continue;
+    }
+    VELOX_CHECK(spiller_->isSpilled(partition));
+    spillPartition(
+        partition, numSpillInputs, spillInputIndicesBuffers_[partition], input);
+  }
+  activeRows_.updateBounds();
+}
+
+void HashBuild::prepareInputIndicesBuffers(
+    vector_size_t numInput,
+    const SpillPartitionNumSet& spillPartitions) {
+  for (const auto& partition : spillPartitions) {
+    if (spillInputIndicesBuffers_[partition] == nullptr) {
+      spillInputIndicesBuffers_[partition] = allocateIndices(numInput, pool());
+    } else {
+      AlignedBuffer::reallocate<vector_size_t>(
```

Comment:
reallocate copies the old data in case it needs to allocate new buffer; since we don't need the old data, it might be more efficient to just check if buffer is large enough and call allocate if not.

Reply:
Good point!